app/pages/1-training.py

```diff
@@ -639,48 +408,61 @@ def training(
     Returns:
         The updated data in the training log storage.
     """
+    global instructor
+
     if page_log_training is None or page_data is None:
         raise PreventUpdate()
 
     if len(page_log_training["loss"]) > page_data["steps"]:
         page_log_training = reset_log()
 
-    instructor = Instructor(
-        main_data["number_qubits"],
-        main_data["number_layers"],
-        n_freqs=page_data["n_freqs"],
-        stepsize=page_data["stepsize"],
-        seed=main_data["seed"],
-        circuit_type=main_data["circuit_type"],
-        data_reupload=main_data["data_reupload"],
-    )
-
-    page_log_training["params"], cost, pred = instructor.step(
-        page_log_training["params"], page_data["noise_params"]
-    )
-
-    page_log_training["loss"].append(cost.item())
-    page_log_training["y_hat"] = pred
+    if (
+        instructor.seed != main_data["seed"]
+        or instructor.model.n_qubits != main_data["number_qubits"]
+        or instructor.model.n_layers != main_data["number_layers"]
+        or instructor.stepsize != page_data["stepsize"]
+        or instructor.n_freqs != page_data["n_freqs"]
+        or instructor.circuit_type != main_data["circuit_type"]
+        or instructor.model.data_reupload != main_data["data_reupload"]
+    ):
+        print("Re-init inst")
+        instructor = Instructor(
+            main_data["number_qubits"],
+            main_data["number_layers"],
+            n_freqs=page_data["n_freqs"],
+            stepsize=page_data["stepsize"],
+            seed=main_data["seed"],
+            circuit_type=main_data["circuit_type"],
+            data_reupload=main_data["data_reupload"],
+        )
     page_log_training["x"] = instructor.x_d
     page_log_training["y"] = instructor.y_d
 
-    data = instructor.calc_hist(
-        params=page_log_training["params"],
-        noise_params=page_data["noise_params"],
-    )
-
-    page_log_training["X"] = np.arange(
-        -len(data) // 2 + 1, len(data) // 2 + 1, 1
-    ).tolist()
-    page_log_training["steps"] = [i for i in range(len(page_log_training["loss"]))]
-    page_log_training["Y"].append(data.tolist())
-
-    if main_data["number_qubits"] > 1:
-        instructor.model.params = page_log_training["params"]
-        ent_cap = instructor.meyer_wallach(
+    try:
+        data = instructor.calc_hist(
+            params=instructor.model.params,
             noise_params=page_data["noise_params"],
         )
+        page_log_training["X"] = np.arange(
+            -len(data) // 2 + 1, len(data) // 2 + 1, 1
+        ).tolist()
 
-        page_log_training["ent_cap"].append(ent_cap)
+        page_log_training["Y"].append(data.tolist())
+
+        if main_data["number_qubits"] > 1:
+            instructor.model.params = instructor.model.params
+            ent_cap = instructor.meyer_wallach(
```

Comment:
I think there might be an issue with qml-essentials here, when setting the n_qubits > 1. I get the following stack trace (when removing the try-catch), when resetting and starting the training (e.g. Ansatz=Circuit19, n_qubits=2, n_layers=2): 
```
Traceback (most recent call last):
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/flask/app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/dash/dash.py", line 1376, in dispatch
    ctx.run(
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/dash/_callback.py", line 507, in add_context
    raise err
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/dash/_callback.py", line 496, in add_context
    output_value = _invoke_callback(func, *func_args, **func_kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/dash/_callback.py", line 43, in _invoke_callback
    return func(*args, **kwargs)  # %% callback invoked %%
           ^^^^^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/app/pages/1-training.py", line 441, in training
    data = instructor.calc_hist(
           ^^^^^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/app/utils/instructor.py", line 166, in calc_hist
    .sample_coefficients(self.model, noise_params=noise_params, cache=False)
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/qml_essentials/coefficients.py", line 30, in sample_coefficients
    coeffs = coefficients(partial_circuit, 1, model.degree)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/pennylane/fourier/coefficients.py", line 178, in coefficients
    return _coefficients_no_filter(f, degree, use_broadcasting)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/pennylane/fourier/coefficients.py", line 262, in _coefficients_no_filter
    f_out = f(sampling_point)
            ^^^^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/qml_essentials/model.py", line 483, in __call__
    return self._forward(
           ^^^^^^^^^^^^^^
  File "<path>/mind-the-qapp/.venv/lib/python3.11/site-packages/qml_essentials/model.py", line 616, in _forward
    if len(result.shape) == 3 and result.shape[0] == 1:
           ^^^^^^^^^^^^
AttributeError: 'list' object has no attribute 'shape'

```
My guess is that this is somehow caused by implicitly setting the execution_type to density when comuting the Meyer-Wallach measure.

Reply:
thanks! already addressed with cirKITers/qml-essentials/pull/61

Reply:
Ah whoops